[Feature] Overview for students' questions

filip-gregurevic opened the pull request:
<!-- Thanks for contributing to Artemis! Before you submit your pull request, please make sure to check the following boxes by putting an x in the [ ] (don't: [x ], [ x], do: [x]) -->
<!-- If your pull request is not ready for review yet, create a draft pull request! -->

### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I added multiple integration tests (Spring) related to the features (with a high test coverage)
- [x] Server: I added `@PreAuthorize` and check the course groups for all new REST Calls (security)
- [x] Server: I implemented the changes with a good performance and prevented too many database calls
- [x] Server: I documented the Java code using JavaDoc style.
- [x] Client: I added multiple integration tests (Jest) related to the features (with a high test coverage)
- [x] Client: I added `authorities` to all new routes and check the course groups for displaying navigation elements (links, buttons)
- [x] Client: I documented the TypeScript code using JSDoc style.
- [x] Client: I added multiple screenshots/screencasts of my UI changes
- [x] Client: I translated all the newly inserted strings into German and English

### Motivation and Context
<!-- Why is this change required? What problem does it solve? -->
<!-- If it fixes an open issue, please link to the issue here. -->
Tutors and instructors are not able to see all questions for a course in one place. In order to see if there are unanswered questions they have to check each exercise and lecture individually.

### Description
<!-- Describe your changes in detail -->
Added an overview for tutors and instructors to see all questions in a course. By clicking the question's ID they can navigate to the question and see the answers.

### Steps for Testing
<!-- Please describe in detail how the reviewer can test your changes. -->

**Setup:** Make sure that you have a test course with some exercises and lectures. Each should have some questions and some answers. Try to check as many combinations as possible.

1. Log in to Artemis
2. Navigate to Course Management
3. Click on the _Q&A_ button
4. Check that all questions are displayed
5. Check that the amounts of answers and approved answers is correct
6. Check that you can navigate to the correct exercise /lecture by clicking the question ID or the title of the exercise / lecture.

### Test Coverage
<!-- Please add the test coverage for all changes files here. You can see this when executing the tests locally (see build.gradle and package.json) or when looking into the corresponding Bamboo build plan -->
* StudentQuestionRessource.java: 84%
* course-questions.component.ts 85%

### Screenshots
<!-- Add screenshots to demonstrate the changes in the UI. -->
<!-- Create a GIF file from a screen recording in a docker container https://toub.es/2017/09/11/high-quality-gif-with-ffmpeg-and-docker/ -->

Button in course view:
<img width="309" alt="Bildschirmfoto 2020-10-08 um 20 57 31" src="https://user-images.githubusercontent.com/51987021/95502230-ca1db680-09a9-11eb-916b-8222a34ba45d.png">

Button in tutor dashboard:
<img width="673" alt="Bildschirmfoto 2020-10-08 um 20 59 24" src="https://user-images.githubusercontent.com/51987021/95502281-e588c180-09a9-11eb-958f-98f5c667caf9.png">

Link in instructor dashboard:
<img width="319" alt="Bildschirmfoto 2020-10-08 um 20 59 33" src="https://user-images.githubusercontent.com/51987021/95502305-ee799300-09a9-11eb-8f5d-a7efb8a3281e.png">



Question overview:
<img width="1430" alt="Bildschirmfoto 2020-09-29 um 21 14 46" src="https://user-images.githubusercontent.com/51987021/94605062-d8c4e900-0298-11eb-904d-16e68c3ead9c.png">

<img width="1410" alt="Bildschirmfoto 2020-09-29 um 21 18 17" src="https://user-images.githubusercontent.com/51987021/94605376-525cd700-0299-11eb-979d-39564017e480.png">

JonasPetry commented on src/main/webapp/app/course/course-questions/course-questions.component.ts:
```diff
@@ -0,0 +1,54 @@
+    courseId: number;
+    studentQuestions: StudentQuestion[];
+
+    paramSub: Subscription;
```
Is there a special reason for this subscription? 

I think `this.courseId = Number(this.route.snapshot.paramMap.get('courseId'));` inside of the ngOnInit would be sufficient.

filip-gregurevic replied:
I've just seen it done with the subscription in quite a few different places, but I like your suggestion better. It's a lot cleaner. Thanks for the suggestion!

JonasPetry commented:
The feature works as you describe it, nice work! However, I have a few suggestions:

- Allow to sort by Answers
- Allow to sort by Approved Answers
- Allow to sort by Exercise/Lecture
- Auto approve answers by tutors/instructors?

What do you think about that? I think the latter one could be done in a follow up.

filip-gregurevic commented:
> The feature works as you describe it, nice work! However, I have a few suggestions:
> 
> * Allow to sort by Answers
> * Allow to sort by Approved Answers
> * Allow to sort by Exercise/Lecture
> * Auto approve answers by tutors/instructors?
> 
> What do you think about that? I think the latter one could be done in a follow up.

@JonasPetry Thanks for reviewing! Could you please elaborate a bit more on the auto approve? I don't quite understand what you mean with that.

JonasPetry commented:
> > The feature works as you describe it, nice work! However, I have a few suggestions:
> > 
> > * Allow to sort by Answers
> > * Allow to sort by Approved Answers
> > * Allow to sort by Exercise/Lecture
> > * Auto approve answers by tutors/instructors?
> > 
> > What do you think about that? I think the latter one could be done in a follow up.
> 
> @JonasPetry Thanks for reviewing! Could you please elaborate a bit more on the auto approve? I don't quite understand what you mean with that.

When I answer a question as a tutor or an instructor, I have to approve my own answer afterwards. I think approving automatically if e.g., an instructor answers a question might improve the user experience.

filip-gregurevic commented:
> > > The feature works as you describe it, nice work! However, I have a few suggestions:
> > > 
> > > * Allow to sort by Answers
> > > * Allow to sort by Approved Answers
> > > * Allow to sort by Exercise/Lecture
> > > * Auto approve answers by tutors/instructors?
> > > 
> > > What do you think about that? I think the latter one could be done in a follow up.
> > 
> > 
> > @JonasPetry Thanks for reviewing! Could you please elaborate a bit more on the auto approve? I don't quite understand what you mean with that.
> 
> @filip-gregurevic When I answer a question as a tutor or an instructor, I have to approve my own answer afterwards. I think approving automatically if e.g., an instructor answers a question, might improve the user experience.

@JonasPetry That makes a lot of sense, I really like it! I'll implement it in a follow-up PR. And now sorting by answers and exercises is also possible btw.

JonasPetry reviewed (approved):
Code looks good now, but I haven't tested the sorting again.

krusche commented:
Can you please add a filter option to hide questions that already have an approved answer?

krusche commented:
It would also be important that instructors / tutors can either directly answer the question or navigate into the corresponding exercise / lecture to answer it

krusche commented:
The number of buttons on the right side is becoming too large. I suggest we find another place for the Q&A button.
Any ideas where we can put so that it still can easily be found?

filip-gregurevic commented:
> Can you please add a filter option to hide questions that already have an approved answer?

@krusche I added a filter option now. See the updated screenshot in the description.

filip-gregurevic commented:
> It would also be important that instructors / tutors can either directly answer the question or navigate into the corresponding exercise / lecture to answer it

@krusche They can already do that by clicking on the id of the question. I can also add a button on the right-hand side if you'd prefer that.

filip-gregurevic commented:
> The number of buttons on the right side is becoming too large. I suggest we find another place for the Q&A button.
> Any ideas where we can put so that it still can easily be found?

@krusche I also thought quite a bit about this. The only good other option, I could think of, is to put it in a tab in the course page with exams, exercises, lectures, but it doesn't feel really intuitive. Also the problem with the buttons only seems to exist for admins and instructors since tutors only have 3 (with this 4) buttons in the course management.

krusche commented:
> > The number of buttons on the right side is becoming too large. I suggest we find another place for the Q&A button.
> > Any ideas where we can put so that it still can easily be found?
> 
> @krusche I also thought quite a bit about this. The only good other option, I could think of, is to put it in a tab in the course page with exams, exercises, lectures, but it doesn't feel really intuitive. Also the problem with the buttons only seems to exist for admins and instructors since tutors only have 3 (with this 4) buttons in the course management.

That's not very obvious. Please put the link also into the right column when you click on the exercise / lecture

stefanwaldhauser reviewed (approved):
Tested it on TS1 and code looks good too

fde312 reviewed (approved):
Works fine and code looks good! Good job! 🚀

MaisiKoleni commented on src/main/webapp/i18n/de/student-question.json:
```diff
@@ -3,7 +3,19 @@
+            "overview": {
+                "title": "Fragen",
+                "noQuestions": "Es wurden noch keine Fragen in diesem Kurs gestellt",
+                "showApproved": "zeige Fragen mit bestätigten Antworten an",
+                "hideApproved": "verstecke Fragen mit bestätigten Antworten"
```
For consideration:
```suggestion
                "showApproved": "Zeige Fragen mit bestätigten Antworten an",
                "hideApproved": "Verstecke Fragen mit bestätigten Antworten"
```

MaisiKoleni commented on src/main/webapp/i18n/en/student-question.json:
```diff
@@ -3,7 +3,19 @@
+            "votes": "Votes",
+            "answers": "Answers",
+            "approvedAnswers": "Approved answers",
+            "exerciseOrLecture": "Exercise/ Lecture",
```
```suggestion
            "exerciseOrLecture": "Exercise / Lecture",
```

MaisiKoleni reviewed (changes requested):
Works well and code looks good to me, I just have some changes/suggestions for the localization.

fde312 reviewed (dismissed):
re-approve

krusche reviewed (changes requested):
The code looks good to me, but I still don't like the big fat button in course management:
1) Please move the button into the course view page (when you click on the course id) **and** add a link in the tutor course dashboard (below complaints, ratings, etc.) so that this is easily accessible for tutors as well **and** add a link in the instructor course dashboard
2) Those buttons should only be visible if the flag `course.studentQuestionsEnabled` is enabled

FrankeLukas commented on src/main/java/de/tum/in/www1/artemis/web/rest/StudentQuestionResource.java:
```diff
@@ -179,7 +184,7 @@ public StudentQuestionResource(StudentQuestionRepository studentQuestionReposito
-     * GET /lectures/{lectureId}/student-questions : get all student questions for exercise.
+     * GET /lectures/{lectureId}/student-questions : get all student questions for lecture.
      * @param lectureId the lecture that the student questions belong to
      * @return the ResponseEntity with status 200 (OK) and with body all student questions for exercise
```
This still says `exercise`, is that correct?

filip-gregurevic replied:
No, I changed it to `lecture` since it is the endpoint to get the questions for a lecture. Must have been forgotten when copy-pasting.

FrankeLukas commented on src/main/java/de/tum/in/www1/artemis/web/rest/StudentQuestionResource.java:
```diff
@@ -200,6 +205,28 @@ public StudentQuestionResource(StudentQuestionRepository studentQuestionReposito
+     *
+     * GET /courses/{courseId}/student-questions : get all student questions for course.
+     * @param courseId the course that the student questions belong to
+     * @return the ResponseEntity with status 200 (OK) and with body all student questions for exercise
```
Same here.

filip-gregurevic replied:
I changed it to say `course` instead of exercise now.

FrankeLukas commented on src/main/java/de/tum/in/www1/artemis/web/rest/StudentQuestionResource.java:
```diff
@@ -200,6 +205,28 @@ public StudentQuestionResource(StudentQuestionRepository studentQuestionReposito
+            return forbidden();
+        }
+        List<StudentQuestion> studentQuestions = studentQuestionService.findStudentQuestionsForCourse(courseId);
+
```
The other similar calls use `hideSensitiveInformation` here. Was that intentionally left out for the course questions?

filip-gregurevic replied:
@FrankeLukas  This is mostly used to hide information students shouldn't be able to see. In this case no such information is sent and the endpoint is only accessible for tutors and instructors.

filip-gregurevic commented:
> The code looks good to me, but I still don't like the big fat button in course management:
> 
> 1. Please move the button into the course view page (when you click on the course id) **and** add a link in the tutor course dashboard (below complaints, ratings, etc.) so that this is easily accessible for tutors as well **and** add a link in the instructor course dashboard
> 2. Those buttons should only be visible if the flag `course.studentQuestionsEnabled` is enabled

@krusche I moved the buttons as you suggested (see updated screenshots in the description) and added the checks.

fde312 reviewed (approved):
Tested the changes and everything works fine. Good job 🚀

stefanwaldhauser commented on src/main/java/de/tum/in/www1/artemis/web/rest/StudentQuestionResource.java:
```diff
@@ -179,7 +184,7 @@ public StudentQuestionResource(StudentQuestionRepository studentQuestionReposito
-     * GET /lectures/{lectureId}/student-questions : get all student questions for exercise.
+     * GET /lectures/{lectureId}/student-questions : get all student questions for lecture.
      * @param lectureId the lecture that the student questions belong to
      * @return the ResponseEntity with status 200 (OK) and with body all student questions for exercise
```
```suggestion
     * @return the ResponseEntity with status 200 (OK) and with body all student questions for lecture
```

stefanwaldhauser reviewed (approved):
Tested the code since the last review and everything still works! Good Job

kloessst commented on src/test/java/de/tum/in/www1/artemis/StudentQuestionIntegrationTest.java:
```diff
@@ -245,4 +245,14 @@ public void editStudentQuestionVotes_asStudent() throws Exception {
+
+    @Test
+    @WithMockUser(username = "tutor1", roles = "TA")
+    public void getAllStudentQuestionsForCourse() throws Exception {
```
```suggestion
    public void testGetAllStudentQuestionsForCourse() throws Exception {
```

FrankeLukas commented on src/main/webapp/i18n/de/student-question.json:
```diff
@@ -3,7 +3,19 @@
-            "deleted": "Frage erfolgreich gelöscht"
+            "deleted": "Frage erfolgreich gelöscht",
+            "text": "Frage",
+            "creationDate": "gestellt am",
```
The English version of this starts upper case, so this should too for consistency.

FrankeLukas commented on src/main/webapp/i18n/de/student-question.json:
```diff
@@ -3,7 +3,19 @@
+            "creationDate": "gestellt am",
+            "votes": "Votes",
+            "answers": "Antworten",
+            "approvedAnswers": "bestätigte Antworten",
```
Same here.

FrankeLukas commented on src/main/webapp/i18n/en/student-question.json:
```diff
@@ -3,7 +3,19 @@
+                "title": "Questions",
+                "noQuestions": "There are no questions for this course",
+                "showApproved": "show questions with approved answers",
+                "hideApproved": "hide questions with approved answers"
```
For both `showApproved` and `hideApproved`: Those should also start with an upper case to be consistent with the German version.

FrankeLukas reviewed (commented):
Lgtm otherwise. :+1:

kloessst reviewed (changes requested):
Code LGTM. Also works fine. 

I just noticed that on clicking on the id of a question with an approved answer I don't jump right to this answer as described in the testing section. I was using Firefox. Is this in the scope of this PR?

![qaOverview](https://user-images.githubusercontent.com/16407766/95864711-12502680-0d66-11eb-8a2e-20623e3ccfd0.gif)

filip-gregurevic commented:
> Code LGTM. Also works fine.
> 
> I just noticed that on clicking on the id of a question with an approved answer I don't jump right to this answer as described in the testing section. I was using Firefox. Is this in the scope of this PR?
> 
> ![qaOverview](https://user-images.githubusercontent.com/16407766/95864711-12502680-0d66-11eb-8a2e-20623e3ccfd0.gif)

@kloessst I didn't phrase that correctly, my bad. For now it is only intended to navigate to the exercise / lecture where the question is. But it would be a nice addition in the future to automatically scroll the the question, I'll keep that in mind.

@FrankeLukas I fixed all the upper-/lower-case issues alsol.

tiffca reviewed (approved):
All works well, nice changes! 

The buttons to go to the page work from the course page, tutor dashboard, and instructor dashboard if questions are enabled. 
The question links correctly go to the respective question. The number of votes, answers, and approved answers are correct and update accordingly. 

tested on test server 1.

FrankeLukas reviewed (approved):
Texts were fixed.

kloessst reviewed (approved):
Code looks good and works fine!
